### DISCUSSION TITLE
Use elpaca for fast build

### DISCUSSION THREAD

**Discussion by hmanhng** [abougouffa/minemacs#139] [Ideas]
Your project amazing
but I noticed straight is really slow
can you try https://github.com/progfolio/elpaca and replace it ?

**Comment by abougouffa** [comment]
Hello @hmanhng ,

Thank you for the feedback and the suggestion.

I've taken a quick look on the project's README and docs and it seems a good candidate, I will check it later.

Thank again!

**Reply by hmanhng** [reply]
I am really looking forward to it!

**Reply by abougouffa** [reply]
@hmanhng 

I'm experimenting on `elpaca` integration in MinEmacs (see [`feat/elpaca`](https://github.com/abougouffa/minemacs/tree/feat/elpaca)). It seems a good alternative, but I'm missing an important feature. **Is there any way to create lock files?**

Reproducibility is a crucial feature for me, so I can have a set of packages versions that I tested, and I know that they will work as expected when I install MinEmacs on another machine.

**Reply by abougouffa** [reply]
I've checked, it seems that `elpaca-write-lockfile` and `elpaca-load-lockfile` do exactly this. I will test it next days, you can test it on the `feat/elpaca` branch.

**Reply by hmanhng** [reply]
Great!
Thank you for your diligence!

**Reply by abougouffa** [reply]
@hmanhng 

I'm making some progress on this, it is very fast compared to `straight.el`. However, I'm facing some issues with it, sometimes it creates dependency loop like this:

![image](https://github.com/abougouffa/minemacs/assets/3716399/7bd3c0de-66dc-4a10-8d3d-66516ec04d4f)

```
    +------ blocked by ------+
    |                        |
    v                        |
 /-----\              /-------------\ 
| magit |            | magit-section |
 \-----/              \-------------/
    |                        ^
    |                        |
    +------ blocked by ------+
```
`magit` is waiting for `magit-section` while `magit-section` is waiting for `magit`

I'm pretty confident that I can find a way to fix this particular case, but if such cases can arrive, it will make MinEmacs unpredictable and unreproducible.

**Reply by hmanhng** [reply]
Add `(elpaca-wait)` below `(use-package magit ... )` maybe that will fix that? ?

**Reply by abougouffa** [reply]
@hmanhng Yep, I know, however, this will make maintaining MinEmacs a pain, as I can't properly test the installation process each time I add a new package! I've opened an issue upstream to see if this can be fixed in `elpaca`.

**Reply by hmanhng** [reply]
@abougouffa I agree with you, sometimes on my old laptop it can cause freezing due to cpu and memory overload when installing too many packages at the same time without control.
Anyway, this is a young project, hopefully the developer will overcome the shortcomings.

**Reply by abougouffa** [reply]
@hmanhng Yep, definitely worth trying, it speeds up the installation process considerably. Thank you for suggesting it, I will keep testing it.

**Reply by abougouffa** [reply]
@hmanhng I've tried to migrate to `elpaca` but I'm getting several issues, and can't have the same behavior on multiple machines.

The project is interesting, but I think it still immature for me to rely on it. I will keep the [`feat/elpaca`](https://github.com/abougouffa/minemacs/tree/feat/elpaca) branch for the moment so I don't forget about it.

In all cases, even if `straight.el` isn't fast enough, the installation is a **one time** process, and then there is no difference between the two.

Thanks again for the suggestion and for your patience!
